### ISSUE TITLE
Normalization doesn't match

### ISSUE THREAD

**Issue by biswajitSM** [tritemio/pycorrelate#2]
* Pycorrelate version: 0.1.0
* Python version: Python 3.6.0 :: Anaconda custom (64-bit)
* Operating System: ubuntu 16.04 lts

### Description
Converted a .pt3 file to Photon-hdf5 with little modification to the phtools. To make sure, the file was properly read, the binned time trace was compared to picoquant-generated-time trace. 

But the auto correlation bu pycorrelate doesn't match with the fcs in picoquant. It seems to have some normalization issue which I couldn't fix. 
![picoquant_fcs](https://user-images.githubusercontent.com/12598051/30520626-ea57e2f4-9bb1-11e7-9fd0-ae52e1eedc16.png)
![pycorrelate_fcs](https://user-images.githubusercontent.com/12598051/30520627-ea5a04f8-9bb1-11e7-8d07-dde94f7bd268.png)

the hdf5 file is attached here:
[Point_A3_075mV(3)_63%_10s_9.hdf5.zip](https://github.com/tritemio/pycorrelate/files/1309054/Point_A3_075mV.3._63._10s_9.hdf5.zip)

```
Paste the command(s) you ran and the output.
If there was a crash, please include the traceback here.
```

**Comment by tritemio** [comment]
@biswajitSM, thanks for the report and data file. 

So far pycorrelate does not include any normalization specific for FCS. I just implemented the raw correlation algorithm and checked that it matches the classical text-book definition of linear correlation (which it does fairly well judging from the last plot in the [example notebook](https://pycorrelate.readthedocs.io/en/latest/notebooks/pycorrelate-examples.html)). 

Adding FCS-specific normalization would be useful. It would also be nice to add an example of fitting  a model using standard python tools (e.g. scipy or lmfit). But I would not have time to work on it for at least 1 month. I would welcome a PR if you want to contribute (and be added as author of course).

**Comment by biswajitSM** [comment]
@tritemio thanks a lot for the quick reply. Indeed the code works much faster. I tried to normalize in a robust way but couldn't find any. The closest i could do:
 G = -1 + G/np.average(G[-10:])
 G: the output of pycorrelate, assuming last 10 values of G approaches a constant value. The result matches quite well but it's prone to error.
Here is how it looks like after normalization;
![selection_054](https://user-images.githubusercontent.com/12598051/30523856-65f5d0f6-9be9-11e7-9653-189f01be69cc.png)
I think it can be done better. i will PR if get a satisfied method. Also I can write some fitting functions.

Thanks
Biswajit

**Comment by tritemio** [comment]
@biswajitSM , I looked into the normalization issue. You can normalize the correlation function with:

```python
def normalize_G(G, t, u, bins):
    duration = max((t.max(), u.max())) - min((t.min(), u.min()))
    Gn = G.copy()
    for i, tau in enumerate(bins[1:]):
        Gn[i] /= (t >= tau).sum() * (u <= (u.max() - tau)).sum()
        Gn[i] *= duration - tau
    return Gn    
```

With your file, this is how I compute the AFC:

```python
import numpy as np
import matplotlib.pyplot as plt
import h5py
import pycollrelate as pyc

fname = './data/Point_A3_075mV(3)_63%_10s_9.hdf5' 
h5 = h5py.File(fname)
t = u = h5['photon_data']['timestamps'][:]
unit = h5['photon_data']['timestamps_specs']['timestamps_unit'][()]

bins = pyc.make_loglags(-7, 0, 10) / unit
G = pyc.pcorrelate(t, u, bins)
Gn = normalize_G(G, t, u, bins)
```

Plotting the results:

```python
fig, ax = plt.subplots(figsize=(10, 6))
plt.plot(bins*unit, np.hstack((Gn[:1], Gn))-1, drawstyle='steps-pre')
plt.xlabel('Time (s)')
plt.grid(True); plt.grid(True, which='minor', lw=0.3)
plt.xscale('log')
plt.xlim(30e-9, 2);
```
<img width="592" alt="download" src="https://user-images.githubusercontent.com/4156237/31506596-b530fce4-af2c-11e7-92f2-71717c492464.png">

Note how the normalization also corrects the slight decrease of the correlation at high time lags.

**Comment by biswajitSM** [comment]
Thanks a lot @tritemio. Indeed, it works perfect.
You are my savior! :D

**Comment by tritemio** [comment]
The function I posted above causes numeric overflows on some platforms/data combinations (it was working on linux machine but failing on windows). The following one works better:

```python
def normalize_G(G, t, u, bins):
    """Normalize ACF and CCF.
    """
    duration = max((t.max(), u.max())) - min((t.min(), u.min()))
    Gn = G.copy()
    for i, tau in enumerate(bins[1:]):
        Gn[i] *= ((duration - tau) 
                  / (float((t >= tau).sum()) * 
                     float((u <= (u.max() - tau)).sum())))
    return Gn
```

This approach was tested in [this publication](https://github.com/tritemio/48-pixel-SPAD-crosstalk-analysis).

**Comment by tritemio** [comment]
Included in pycorrelate v0.3
